### ISSUE TITLE
crypt base32768 file names encoding on case insensitive remotes  (e.g. OneDrive) is not 100% safe

### ISSUE THREAD

**Issue by kapitainsky** [rclone/rclone#6803]
For various reasons I was trying to understand base32768 in details and I have suspicion that base32768 encoding rclone is using is not 100% safe on case insensitive remotes like OneDrive.

The reason is that so called safe alphabet used contains both upper and lower case characters.

I have raised the question with @Max-Sum as his go encoding procedure is used in rclone.

https://github.com/Max-Sum/base32768/issues/1

Chances of upper/lower case collisions are extremely low (but not zero) given that 99% of alphabet used does not have  upper and lower cases (asian characters) and it is only used for file names. We would need by chance two different files resulting with base32768 encoding results like:

䏠Ҡ䈠䉀䉠䊀䊠䋀䋠䌀䌠䍀䍠䎀䎠䏀
and
䏠ҡ䈠䉀䉠䊀䊠䋀䋠䌀䌠䍀䍠䎀䎠䏀

In this example Ҡ and ҡ are problematic and as I understand it they are both part of alphabet used by rclone.

Maybe I am wrong but I think it would be helpful for somebody more knowledgeable to double check it.

**Comment by ncw** [comment]
I guess we should work out what the probability of a collision is.

I wrote a little program to figure out how many characters are case folding

https://go.dev/play/p/IEOcELc1yrH

```
Case folded ƀ into Ƀ
Case folded ɀ into Ɀ
Case folded ɠ into Ɠ
Case folded ʀ into Ʀ
Case folded ҡ into Ҡ
Case folded ԁ into Ԁ
Case folded ⴀ into Ⴀ
Case folded ꭰ into Ꭰ
Case folded ꮐ into Ꮐ
Case folded ⰰ into Ⰰ
Case folded ⲁ into Ⲁ
Case folded ⲡ into Ⲡ
Case folded ⳁ into Ⳁ
Case folded ⴀ into Ⴀ
Case folded ꙁ into Ꙁ
Case folded ꝁ into Ꝁ
Case folded ꞁ into Ꞁ
Found 17 case folding characters
```

Note that 16 of these characters only contain either the upper case or lower case character which means that when rclone is generating file names, they will never contain something which can be confused by case folding.

The only problem character is `ⴀ` and `Ⴀ` which does contain both an upper case and lower case version.

All the cloud providers preserve the case of file names even if they are case insensitive so none of this will be a problem except if we generate two identical file names except for `ⴀ` and `Ⴀ`

The minimum size of an encrypted file name is 16 bytes (unencoded) so 128 bits which makes 9 base32768 characters.

The probability of no `ⴀ` or `Ⴀ` characters is (32766/32768)^9 = 0.9994508176851031, so only 1 in 1820 file names will have a `ⴀ` or `Ⴀ` (roughly).

The probability that we have an identical file name, with only a differing case of the `ⴀ` or `Ⴀ` is approx 1 in 2**128 assuming the encryption generates uniformly distributed binary digits (which it should).

So I think it is vanishingly unlikely that this will ever be a problem.

@Max-Sum what do you think?

**Comment by kapitainsky** [comment]
> I guess we should work out what the probability of a collision is.
> 
> I wrote a little program to figure out how many characters are case folding
> 
> https://go.dev/play/p/IEOcELc1yrH
> 
> ```
> Case folded ƀ into Ƀ
> Case folded ɀ into Ɀ
> Case folded ɠ into Ɠ
> Case folded ʀ into Ʀ
> Case folded ҡ into Ҡ
> Case folded ԁ into Ԁ
> Case folded ⴀ into Ⴀ
> Case folded ꭰ into Ꭰ
> Case folded ꮐ into Ꮐ
> Case folded ⰰ into Ⰰ
> Case folded ⲁ into Ⲁ
> Case folded ⲡ into Ⲡ
> Case folded ⳁ into Ⳁ
> Case folded ⴀ into Ⴀ
> Case folded ꙁ into Ꙁ
> Case folded ꝁ into Ꝁ
> Case folded ꞁ into Ꞁ
> Found 17 case folding characters
> ```


For every character you test from safe alphabet you also have to test consecutive 31 characters - so 544 in total.

Also we do not have any guarantee that if safe alphabet initial letter does not have cases there are no hidden in next 31 characters.

**Comment by kapitainsky** [comment]
For my example with Ҡ character all below characters are used in encoding:

![image](https://user-images.githubusercontent.com/15635386/222711646-7eddfa79-ef17-44f1-abad-44aa88aa92e8.png)

16 pairs of upper and lower case

**Comment by ncw** [comment]
I corrected my program https://go.dev/play/p/SK5G4dnHM6T

The result was

Found 521 case folding and 199 duplicate case folding characters out of 32896

I don't think the 521 is a problem

However the 199 are a potential problem...

However it is still a 1 in 2**128 chance that you have a matching file name which differs by only one of the case folding characters.

Since the case folding characters are relatively rare (0.6%) it is unlikely that we will have more than one in a file name.

> Also I believe that problem is correctable and case insensitive safe alphabet can be generated. Obvious approach would be to avoid any upper(Lu) or lower (LI) and stick with other letters (Lo) and symbols etc.

We can't change the existing alphabet - this would break existing crypt mounts. We'd have to make a new alphabet with an option to use it.

**Comment by kapitainsky** [comment]
> However the 199 are a potential problem...

Correct - case folding characters are not a problem. The issue are duplicate cases.

Even without math it is obvious that chances of collision are small - however my point is that it is avoidable..

**Comment by kapitainsky** [comment]
> We can't change the existing alphabet - this would break existing crypt mounts. We'd have to make a new alphabet with an option to use it.

Yes this would be sweet. To have new option base32768 case insensitive.

I will think myself how to do it

Here there is script we could use for this job:

https://github.com/qntm/base32768/blob/main/scripts/gen.js

I will try to look into it.

**Comment by kapitainsky** [comment]
I did some testing using https://github.com/qntm/base32768/blob/main/scripts/gen.js and it is promising.

With latest unicode(15.0.0)

removing both Lu and Ll characters is not good as only 1022 ranges are possible (we need 1028)

removing ranges with only Ll - 1026  - not good

but removing only Lu gives us 1030.

Problem - we only have enough case insensitive ranges for unicode version 14 (from 2021 September 14) or newer

Question is which version OneDrive is using????

**Comment by kapitainsky** [comment]
few tests later....

"Question is which version OneDrive is using????" I think it does not matter... below some latest codes introduced by Unicode 14.0.0:

ࡷࡷࢊࢊࡰࡰ
(some new arabic characters) Most likely your browser like mine does not know them yet. But then simply displays then as code points.

The same happens with Onedrive - I can create filenames with all new characters. Just do not see good looking glyphs but bare code points. 

So I am positive that new case insensitive alphabet can be created. I will look more into it during weekend.




ࢌࢌ

**Comment by ncw** [comment]
Yes, this needs to go upstream to @Max-Sum first

**Comment by kapitainsky** [comment]
done

https://github.com/Max-Sum/base32768/pull/2

Up to @Max-Sum now

**Comment by Max-Sum** [comment]
> Yes, this needs to go upstream to @Max-Sum first

Merged. Cheers @kapitainsky

**Comment by kapitainsky** [comment]
I have added PR for rclone

https://github.com/rclone/rclone/pull/6806

now up to @ncw to include it

**Comment by kapitainsky** [comment]
> I have made an upstream issue about this here [qntm/base32768#21](https://github.com/qntm/base32768/issues/21) on the creator of the base32768 encoding's github - I'd like their feedback.
> 
> I also tried harder to calculate the probability of this happening, and I think you'd have to put around 10²¹ files in a single directory before you got a collision, so I don't think it is very likely to happen ever.

Probability is also low for base64. And yes it is very small so this issue is not urgent.
But as solution exist and I heavily use OneDrive I will work on making it better:)

**Comment by ncw** [comment]
:-)  Qntm requested that if we do make a new encoding we don't call it base32768 so we don't confuse people.

**Comment by kapitainsky** [comment]
As qntm created base32768 and did very valuable research on safe characters I fully agree to respect his wish.

For now there is nothing to worry about as there is nothing to be named. But I hope to have soon super safe case insensitive solution with zero folding and duplicated characters.  Then will think how to call it:)

**Comment by major-mayer** [comment]
Is it safe to use the new "base32768"  encoding for OneDrive at the moment?
It seems to have fixed my too long file names, but as this https://github.com/rclone/rclone/pull/6806 has not been merged, I can't use the new "base32768ci" option atm.
When the new case-insensitive option gets merged, will I have to re-upload all my files again?

**Comment by kapitainsky** [comment]
> Is it safe to use the new "base32768" encoding for OneDrive at the moment?

Yes it is "safe" - in similar way like using hashes is safe to check if two files are identical. There is theoretical chance of collision but in real life it can be ignored. You have bigger chance of winning lottery. ..

@ncw did some math here:

>I also tried harder to calculate the probability of this happening, and I think you'd have to put around 10²¹ files in a single directory before you got a collision, so I don't think it is very likely to happen ever.

Now maybe one day `base32768ci` will appear... I know what to do but can not work on this now.

> When the new case-insensitive option gets merged, will I have to re-upload all my files again?

No. encoding alphabet can be changed server-side

**Comment by ncw** [comment]
I am going to close this as I think this issue, while real, will never happen in practice. If it does I shall re-open!
